org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/outline/ModelBuilder.scala

```diff
@@ -0,0 +1,242 @@
+            renderTuple(sb, args)
+            if (dp)
+              sb.append(")")
+          case "_root_.scala.<byname>" =>
+            sb.append("=> ")
+            renderType(sb, args.head, List(), dp)
+          case "_root_.scala.<repeated>" =>
+            renderType(sb, args.head, List(), dp)
+            sb.append("*")
+          case _ =>
+            renderType(sb, tpt, args)
+        }
+      }
+      def renderType(sb: StringBuilder, tt: Tree, args: List[Tree] = List(), dp: Boolean = false): Unit = {
+        tt match {
+          case AppliedTypeTree(tpt: Tree, args: List[Tree]) => renderATT(sb, tt, tpt, args, dp)
+          case _ =>
+            sb.append(tt.toString().split("\\.").reverse.head)
+            if (!args.isEmpty) {
+              sb.append("[")
+              args.foreach(a => { renderType(sb, a, List(), dp); sb.append(", ") })
+              sb.setLength(sb.length - 2)
+              sb.append("]")
+            }
+        }
+      }
+      def showType(tt: Tree): String = {
+        val sb = new StringBuilder
+        renderType(sb, tt)
+        sb.toString() //.split("\\.").reverse.head
+      }
+      def showTypeList(tl: List[TypeDef]): String = {
+        val sb = new StringBuilder
+        renderTypeList(sb, tl)
+        sb.toString()
+      }
+      def renderTypeList(sb: StringBuilder, tl: List[TypeDef]): Unit = {
+        if (!tl.isEmpty) {
+          sb.append("[")
+          tl.foreach(td => { sb.append(td.name); renderTypeList(sb, td.tparams); sb.append(", ") })
+          sb.setLength(sb.length - 2)
+          sb.append("]")
+        }
+      }
+      t match {
+        case Template(_, _, _) => t.children.foreach(x => updateTree(parent, x))
+        case PackageDef(pid, stats) => {
+          if (pid.pos.isDefined && pid.pos.start != pid.pos.end) {
+            val ch = PackageNode(pid.toString(), parent)
+            setPos(ch, t)
+            parent.addChild(ch)
+            //t.children.foreach(x => updateTree(parent, x))
+          }
+          t.children.foreach(x => updateTree(parent, x))
+        }
+
+        case ClassDef(mods, name, tpars, templ) => {
+          val ch = ClassNode(name.decodedName.toString(), parent, showTypeList(tpars))
+          setPos(ch, t)
+          parent.addChild(ch)
+          ch.setFlags(mods.flags)
+          t.children.foreach(x => updateTree(ch, x))
+        }
+
+        case `noSelfType` =>
+
+        case ValDef(mods, name, tpt, rsh) =>
+          if ((mods.flags & SYNTHETIC) == 0) {
+            val ch = if ((mods.flags & MUTABLE) == 0)
+              ValNode(name.decodedName.toString(), parent, if (tpt.isEmpty) None else Some(showType(tpt)))
+            else
+              VarNode(name.decodedName.toString(), parent, if (tpt.isEmpty) None else Some(showType(tpt)))
+            setPos(ch, t)
+            ch.setFlags(mods.flags)
+            parent.addChild(ch)
+
+          }
+
+        case DefDef(mods, name, tparamss, vparamss, tpt, rsh) =>
```

Comment:
I don't see recursing inside the right-hand side of this method. It would be really useful to show local methods (the current outline view does that).